feat: Laravel 11 support, fixes #5962

stasadev opened the pull request:
## The Issue

- #5962

The PHP version is set to 8.1 for Laravel in DDEV HEAD, so it cannot pull Laravel 11, and it pulls Laravel 10.

Laravel 11 introduced new skeleton, and now it has a new [`mariadb`](https://github.com/laravel/laravel/blob/6ea57d766ffc7948adbf02b6ac63670d43580316/config/database.php#L62-L80) database driver.

## How This PR Solves The Issue

- Uses `mariadb` driver by default.
- Sets PHP version to 8.2, including docs.
- Fix Mailpit (Laravel 11 sends emails to log file by default)
- Laravel 11 adds [commented](https://github.com/laravel/laravel/blob/6ea57d766ffc7948adbf02b6ac63670d43580316/.env.example#L22-L27) variables for database, like this:

```dotenv
DB_CONNECTION=sqlite
# DB_HOST=127.0.0.1
# DB_PORT=3306
# DB_DATABASE=laravel
# DB_USERNAME=root
# DB_PASSWORD=
```

And DDEV doesn't replace the commented lines with new values (it adds `DB_` variables at the end of `.env`), I fixed it.

## Manual Testing Instructions

Use the quickstart for Laravel https://ddev--5963.org.readthedocs.build/en/5963/users/quickstart/#laravel

Check `.env` file.

Check Mailpit:
```
ddev mailpit
ddev php artisan tinker --execute='Mail::raw("Hello World!", function($msg) { $msg->to("test@example.com")->subject("Test Email"); });'
```

Test it with:
```
ddev config --database=mysql:5.7
ddev config --database=mysql:8.0
ddev config --database=postgres:15
```

Check SQLite config from the quickstart.

Test older Laravel:
```
ddev composer create --prefer-dist laravel/laravel:^10
ddev php artisan migrate
```

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

## Related Issue Link(s)

- https://github.com/laravel/laravel/pull/6367

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->

rfay reviewed (commented):
Awesome! This will need work on https://github.com/ddev/test-laravel of course.

tyler36 commented:
> By default, new Laravel applications use SQLite for database storage

https://laravel.com/docs/11.x/releases#application-defaults

My `.env` is set to "mariadb" when using this PR.

```env
DB_CONNECTION="mariadb"
DB_HOST="db"
DB_PORT="3306"
DB_DATABASE="db"
DB_USERNAME="db"
DB_PASSWORD="db"
```

Personally, I prefer to be using a "real" database because it more closely mimics production environments. 
I would vote for _not_ setting DB to sqlite. Perhaps we need to add a comment as appropriate?
- "DDEV sets a default mariadb database to better represent a production environment. Visit -here- to learn about change database types."
- "DDEV follows Laravel defaults and sets Laravel 11 projects to a sqlite database. Visit -here- to learn about change database types."

stasadev commented:
I've updated the PR with new fixes for Mailpit, returned `DB_CONNECTION=mysql` when using `ddev config --database=mysql:8.0` and added a new tip for SQLite, thanks @tyler36 !

rfay commented on docs/content/users/quickstart.md:
```diff
@@ -323,7 +323,7 @@ The Laravel project type can be used for [Lumen](https://lumen.laravel.com/) lik
     cd my-laravel-app
-    ddev config --project-type=laravel --docroot=public
+    ddev config --project-type=laravel --docroot=public --php-version=8.2
     ddev composer create --prefer-dist laravel/laravel -y
```
I think we should start to show our expectations on these quickstarts, show the tag they've been tested with. Also `-y` is no longer required with composer create.

```suggestion
    ddev composer create --prefer-dist laravel/laravel:^11
```

rfay replied:
I was accidentally testing with a `ddev` that wasn't for this version, and got this during composer install:
`SQLSTATE[HY000]: General error: 1273 Unknown collation: 'utf8mb4_0900_ai_ci' (Connection: mysql, SQL: select table_name as `name`, (data_length + index_length) as `size`, table_comment as `comment`, engine as `engine`, table_collation as `collation` from information_schema.tables where table_schema = 'db' and table_type in ('BASE TABLE', 'SYSTEM VERSIONED') order by table_name).`

They seem to default to `mysql:8.0`? 
This is fixed by using `--database=mysql:8.0`. Their system requirements certainly don't hint at this problem, but I'm sure you understand it.

But it means that this quickstart is dependent on using DDEV v1.23.0+, I wish it wasn't.  Maybe we need a tip about it?

stasadev replied:
> Unknown collation: 'utf8mb4_0900_ai_ci'

This is because DDEV sets `DB_CONNECTION=mysql`, but uses MariaDB.

https://github.com/laravel/laravel/blob/11.x/config/database.php#L42-L80

This means when you have  `DB_CONNECTION=mariadb` and MariaDB - it will work fine, but because DDEV v1.22.7 sets it to `mysql`, it complains about a collation.

> But it means that this quickstart is dependent on using DDEV v1.23.0+, I wish it wasn't. Maybe we need a tip about it?

It is already dependent on v1.23.0+ because of

- #5058

But I am open to discussion here (note that I made an updated quickstart for Laravel at the top of the page in #5962).

rfay commented on docs/content/users/quickstart.md:
```diff
@@ -323,7 +323,7 @@ The Laravel project type can be used for [Lumen](https://lumen.laravel.com/) lik
     mkdir my-laravel-app
     cd my-laravel-app
-    ddev config --project-type=laravel --docroot=public
+    ddev config --project-type=laravel --docroot=public --php-version=8.2
```
Just a note that if using an earlier version of DDEV `mysql:8.0` seems to be required:

```
    ddev config --project-type=laravel --docroot=public --php-version=8.2 --database=mysql:8.0
```

rfay replied:
I think that using `--php-version=8.2` is fine, but it makes me wonder whether we should show explicit defaults then 1) it's already the DDEV default and 2) we're already overriding it in the Laravel handling in DDEV.

stasadev replied:
I believe this is good when people use the latest docs without even noticing it.

rfay commented on docs/content/users/quickstart.md:
````diff
@@ -333,20 +333,28 @@ The Laravel project type can be used for [Lumen](https://lumen.laravel.com/) lik
     ```
 
+!!!tip "How to use SQLite database for Laravel?"
+    DDEV sets a default MariaDB database to better represent a production environment.
````
```suggestion
    DDEV defaults to using a MariaDB database to better represent a production environment.
```

rfay commented on docs/content/users/quickstart.md:
````diff
@@ -333,20 +333,28 @@ The Laravel project type can be used for [Lumen](https://lumen.laravel.com/) lik
     ddev launch
     ```
 
+!!!tip "How to use SQLite database for Laravel?"
````
```suggestion
!!!tip "Want to use a SQLite database for Laravel?"
```

rfay commented on .spellcheckwordlist.txt:
```diff
@@ -574,7 +574,7 @@ socio
 somefile
 sql
-sqlite
+SQLite
```
Such a tiny nit, please move this to the capitalized section earlier.

rfay commented on docs/content/users/quickstart.md:
```diff
@@ -333,20 +333,28 @@ The Laravel project type can be used for [Lumen](https://lumen.laravel.com/) lik
     Normal details of a Composer build for Magento 2 are on the [Magento 2 site](https://experienceleague.adobe.com/docs/commerce-operations/installation-guide/composer.html). You must have a public and private key to install from Magento’s repository. When prompted for “username” and “password” in `composer create`, it’s asking for your public key as "username" and private key as "password".
 
-    Note that you can install the Adobe/Magento composer credentials in your global `~/.ddev/homeadditions/.composer/auth.json` and never have to find them again. See [In-Container Home Directory and Shell Configuration](../extend/in-container-configuration).
+    Note that you can install the Adobe/Magento composer credentials in your global `~/.ddev/homeadditions/.composer/auth.json` and never have to find them again. See [In-Container Home Directory and Shell Configuration](extend/in-container-configuration.md).
```
Hmm, the other link is working fine on latest docs. https://ddev.readthedocs.io/en/latest/users/quickstart/#magento

This link works too! https://ddev--5963.org.readthedocs.build/en/5963/users/quickstart/#magento

stasadev replied:
Yeah, I just saw a complain from local mkdocs, and fixed it.

rfay commented on pkg/ddevapp/envfile.go:
```diff
@@ -25,9 +26,9 @@ func WriteProjectEnvFile(envFilePath string, envMap map[string]string, envText s
 		// If the item is already in envText, use regex to replace it
 		// otherwise, append it to the envText.
-		exp := regexp.MustCompile(fmt.Sprintf(`((^|[\r\n]+)%s)=(.*)`, k))
+		exp := regexp.MustCompile(fmt.Sprintf(`(^|[\r\n]+)#*\s*(%s)=(.*)`, k))
```
Please add a comment explaining the purpose of this regex. You apparently had to study it!

stasadev replied:
Done.

rfay commented on pkg/ddevapp/laravel.go:
```diff
@@ -67,8 +71,8 @@ func laravelPostStartAction(app *DdevApp) error {
+// laravelConfigOverrideAction overrides php_version for Laravel, requires PHP8.2
 func laravelConfigOverrideAction(app *DdevApp) error {
-	app.PHPVersion = nodeps.PHP81
+	app.PHPVersion = nodeps.PHP82
```
Same conversation as above. This is fine, but we need to figure out a more general strategy about whether we override things that are already the default

stasadev replied:
I saw that `laravelConfigOverrideAction` was added when Laravel 10 required PHP 8.1.
I decided to keep it because it will be easier to override it in the future.

rfay reviewed (changes requested):
This is a great piece of work and much appreciated. Minor comments. I'd love to figure out how the quickstart could work without the logic added by this PR.

stasadev commented:
Related:

- https://github.com/laravel/laravel/pull/6367

The collation is reverted to its previous value in `laravel/laravel` 11.0.2 for MariaDB https://github.com/laravel/laravel/commit/79969c99c6456a6d6edfbe78d241575fe1f65594

Maybe MySQL will also be reverted :thinking:

Let's see what happens.

rfay commented:
I see a million people had the problem I had (using DDEV without this PR). https://github.com/laravel/laravel/pull/6367

stasadev commented:
> Maybe MySQL will also be reverted 🤔

I don't think they will revert it. I applied a [recommendation](https://github.com/laravel/laravel/pull/6369#issuecomment-1996875154) from Laravel Core team for MySQL 5.7 to `laravelPostStartAction`.

---

And I noticed that with this new change to `mariadb` all existing Laravel 10 (and older) projects would be broken:

```
ddev composer create --prefer-dist laravel/laravel:^10
ddev php artisan migrate
// InvalidArgumentException Database connection [mariadb] not configured
```

I know that we can recommend:

```
ddev config --disable-settings-management=true
```

But we can easily check if `mariadb` is available in the Laravel config, so I fixed it.

---

I think I've covered all possible scenarios. Ready to review.

davidoskay commented:
@stasadev 
```
// Laravel team recommendation for MySQL 5.7 is to use "mariadb" driver
// https://github.com/laravel/laravel/pull/6369#issuecomment-1996875154
```
This is one of two options. Setting mariadb for now only changes this collation after this  https://github.com/laravel/laravel/pull/6367 

Your other option is to stay with mysql driver and add `DB_COLLACTION=utf8mb4_unicode_ci` in your .env (laravel11+) as the same does setting mariadb connection for now.

rfay commented:
That's pretty weird, with MySQL 5.7 being EOL.

stasadev commented:
> This is one of two options.

@davidoskay I applied the other one with `DB_CONNECTION=mariadb` only because it was written from the Laravel Core team member, and it doesn't require me to check if `DB_COLLATION` env exists.

stasadev commented:
> That's pretty weird, with MySQL 5.7 being EOL.

Yes, pretty weird. And MySQL 5.7 is "supported" in Laravel 11 https://laravel.com/docs/11.x/database#introduction

stasadev commented:
And finally, the changes to MariaDB and MySQL for collation have been reverted in Laravel 11.0.3.

rfay reviewed (approved):
It all seems to be great, and now that Laravel has sorted things, the quickstart works with most any DDEV version.
